prow/cmd/peribolos/main.go

```diff
@@ -244,34 +291,266 @@ func configureOrgMembers(opt options, client configureOrgMembersClient, orgName
+func validateTeamNames(orgConfig org.Config) error {
+	// Does the config duplicate any team names?
+	used := sets.String{}
+	dups := sets.String{}
+	for name, orgTeam := range orgConfig.Teams {
+		if used.Has(name) {
+			dups.Insert(name)
+		} else {
+			used.Insert(name)
+		}
+		for _, n := range orgTeam.Previously {
+			if used.Has(n) {
+				dups.Insert(n)
+			} else {
+				used.Insert(n)
+			}
+		}
+	}
+	if n := len(dups); n > 0 {
+		return fmt.Errorf("%d duplicated names: %s", n, strings.Join(dups.List(), ", "))
+	}
+	return nil
+}
+
+type teamClient interface {
+	ListTeams(org string) ([]github.Team, error)
+	CreateTeam(org string, team github.Team) (*github.Team, error)
+	DeleteTeam(id int) error
+}
+
+// configureTeams returns the ids for all expected team names, creating/deleting teams as necessary.
+func configureTeams(client teamClient, orgName string, orgConfig org.Config) (map[string]github.Team, error) {
+	if err := validateTeamNames(orgConfig); err != nil {
+		return nil, err
+	}
+
+	// What teams exist?
+	ids := map[int]github.Team{}
+	ints := sets.Int{}
+	teamList, err := client.ListTeams(orgName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list teams: %v", err)
+	}
+	for _, t := range teamList {
+		ids[t.ID] = t
+		ints.Insert(t.ID)
+	}
+
+	// What is the lowest ID for each team?
+	older := map[string][]github.Team{}
+	names := map[string]github.Team{}
+	for _, t := range ids {
+		n := t.Name
+		switch val, ok := names[n]; {
+		case !ok: // first occurrence of the name
+			names[n] = t
+		case ok && t.ID < val.ID: // t has the lower ID, replace and send current to older set
+			names[n] = t
+			older[n] = append(older[n], val)
+		default: // t does not have smallest id, add it to older set
+			older[n] = append(older[n], val)
+		}
+	}
+
+	// What team are we using for each configured name, and which names are missing?
+	matches := map[string]github.Team{}
+	missing := map[string]org.Team{}
+	used := sets.Int{}
+	for name, orgTeam := range orgConfig.Teams {
+		t := findTeam(names, name, orgTeam.Previously...)
+		if t == nil {
+			missing[name] = orgTeam
+			continue
+		}
+		matches[name] = *t // t.Name != name if we matched on orgTeam.Previously
+		used.Insert(t.ID)
+	}
+
+	// Create any missing team names
+	var failures []string
+	for name, orgTeam := range missing {
+		t := &github.Team{Name: name}
+		if orgTeam.Description != nil {
+			t.Description = *orgTeam.Description
+		}
+		if orgTeam.Privacy != nil {
+			t.Privacy = string(*orgTeam.Privacy)
+		}
+		t, err := client.CreateTeam(orgName, *t)
+		if err != nil {
+			logrus.WithError(err).Warnf("Failed to create %s in %s", name, orgName)
+			failures = append(failures, name)
+			continue
+		}
+		matches[name] = *t
+		used.Insert(t.ID)
+	}
+	if n := len(failures); n > 0 {
+		return nil, fmt.Errorf("failed to create %d teams: %s", n, strings.Join(failures, ", "))
+	}
+
+	// Delete undeclared teams.
```

Comment:
Should we check the team deletion against a delta too?

Reply:
Mind if I handle that in a follow-up PR?

Reply:
Not at all. Could we leave a TODO comment as a breadcrumb?

Reply:
Done